File: mizar/cni/mizarcni.py
# SPDX-License-Identifier: MIT
# Copyright (c) 2020 The Authors.

# Authors: Sherif Abdelwahab <@zasherif>
#          Phu Tran          <@phudtran>

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:The above copyright
# notice and this permission notice shall be included in all copies or
# substantial portions of the Software.THE SOFTWARE IS PROVIDED "AS IS",
# WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
# TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
# FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR
# THE USE OR OTHER DEALINGS IN THE SOFTWARE.

import subprocess
import sys
import logging
from logging.handlers import SysLogHandler
import json
import os
import fs
from mizar.common.common import *
from mizar.daemon.interface_service import InterfaceServiceClient
from mizar.proto.interface_pb2 import *
from pyroute2 import IPRoute, NetNS

logger = logging.getLogger('mizarcni')
handler = logging.FileHandler('/tmp/mizarcni.log')
logger.addHandler(handler)
logger.setLevel(logging.INFO)


class Cni:
    def __init__(self):
        stdin = ''.join(sys.stdin.readlines())
        self.command = os.environ.get("CNI_COMMAND")  # ADD | DEL | VERSION
        self.container_id = os.environ.get("CNI_CONTAINERID")
        self.netns = os.environ.get("CNI_NETNS")
        self.interface = os.environ.get("CNI_IFNAME")
        self.cni_path = os.environ.get("CNI_PATH")
        self.cni_args = os.environ.get("CNI_ARGS")
        self.cni_args_dict = {}
        netns_folder = "/var/run/netns/"
        if not self.netns.startswith(netns_folder):
            dst_netns = self.netns.replace('/', '_')
            dst_netns_path = os.path.join(netns_folder, dst_netns)
            logger.info("hochan netns:{} dst_netns_path:{}".format(self.netns, dst_netns_path))
            if self.command == "ADD":
                errorcode = bindmount_netns(self.netns, dst_netns_path)
                if errorcode != 0:
                    logger.info("failed to bind mount {} to {}: error code {}".format(
                        self.netns, dst_netns_path, errorcode))
                    raise OSError("failed to bind mount netns {} to {}, error code: {}".format(
                        self.netns, dst_netns_path, errorcode))
            self.netns = dst_netns_path
        config_json = json.loads(stdin)

        # expected parameters in the CNI specification:
        self.cni_version = config_json.get("cniVersion")
        self.network_name = config_json.get("name")
        self.plugin = config_json.get("type")

        if len(self.cni_args) > 1:
            self.cni_args_dict = dict(i.split("=")
                                      for i in self.cni_args.split(";"))
            self.k8s_namespace = self.cni_args_dict.get(
                'K8S_POD_NAMESPACE', '')
            self.k8s_pod_name = self.cni_args_dict.get('K8S_POD_NAME', '')

            # TODO: parse 'Arktos specific' CNI_ARGS
            self.k8s_pod_tenant = self.cni_args_dict.get('K8S_POD_TENANT', '')

            self.pod_id = PodId(
                k8s_pod_name=self.k8s_pod_name,
                k8s_namespace=self.k8s_namespace,
                k8s_pod_tenant=self.k8s_pod_tenant
            )

            self.interface_id = InterfaceId(
                pod_id=self.pod_id, interface=self.interface)

        self.iproute = IPRoute()

    def __del__(self):
        if self.command != "VERSION":
            logger.info("Closing IPRoute")
            self.iproute.close()

    def run(self):
        if len(self.cni_args_dict) != 0:
            logging.info("CNI ARGS {}".format(self.cni_args_dict))
        val = "Unsuported cni command!"
        switcher = {
            'ADD': self.do_add,
            'DEL': self.do_delete,
            'GET': self.do_get,
            'VERSION': self.do_version
        }

        func = switcher.get(self.command, lambda: "Unsupported cni command")
        if func:
            func()
        print(val)
        exit(1)

    def do_add(self):

        # Construct a CniParameters grpc message
        param = CniParameters(pod_id=self.pod_id,
                              netns=self.netns,
                              interface=self.interface)
        logger.info("Doing CNI add for {}".format(self.pod_id))
        # Consume new (and existing) interfaces for this Pod
        interfaces = InterfaceServiceClient(
            "localhost").ConsumeInterfaces(param).interfaces

        if len(interfaces) == 0:
            logger.error("No interfaces found for {}".format(self.pod_id))
            exit(1)

        result = {
            "cniVersion": self.cni_version,
            "interfaces": [],
            "ips": []
        }

        # Construct the result string
        idx = 0
        for interface in interfaces:
            self.activate_interface(interface)
            itf = {
                "name": interface.interface_id.interface,
                "mac": interface.address.mac,
                "sandbox": self.netns
            }
            ip = {
                "version": interface.address.version,
                "address": "{}/{}".format(interface.address.ip_address, interface.address.ip_prefix),
                "gateway": interface.address.gateway_ip,
                "interface": idx
            }
            result['interfaces'].append(itf)
            result['ips'].append(ip)
            idx += 1

        print(json.dumps(result))
        exit(0)

    def activate_interface(self, interface):
        """
        moves the interface to the CNI netnt, rename it, set the IP address, and
        the GW.
        """

        _, netns = os.path.split(self.netns)
        iproute_ns = NetNS(netns)
        veth_index = get_iface_index(interface.veth.name, self.iproute)

        actived_idxs = iproute_ns.link_lookup(operstate="UP")
        if (veth_index in actived_idxs):
            return

        logger.info("Move interface {}/{} to netns {}".format(
            interface.veth.name, veth_index, netns))
        self.iproute.link('set', index=veth_index, net_ns_fd=netns)

        # configure and activate interfaces inside the netns
        iproute_ns.link('set', index=veth_index, ifname=self.interface)
        lo_idx = iproute_ns.link_lookup(ifname="lo")[0]
        iproute_ns.link('set', index=lo_idx, state='up')

        iproute_ns.link('set', index=veth_index, state='up')

        iproute_ns.addr('add', index=veth_index, address=interface.address.ip_address,
                        prefixlen=int(interface.address.ip_prefix))
        iproute_ns.route('add', gateway=interface.address.gateway_ip)

        # Disable TSO and checksum offload as xdp currently does not support
        logger.info("Disable tso for pod")
        cmd = "ip netns exec {} ethtool -K {} tso off gso off ufo off".format(
            netns, "eth0")
        rc, text = run_cmd(cmd)
        logger.info("Executed cmd {} tso rc: {} text {}".format(cmd, rc, text))
        logger.info("Disable rx tx offload for pod")
        cmd = "ip netns exec {} ethtool --offload {} rx off tx off".format(
            netns, "eth0")
        rc, text = run_cmd(cmd)
        logger.info("Executed cmd {} rc: {} text {}".format(cmd, rc, text))

    def do_delete(self):
        param = CniParameters(pod_id=self.pod_id,
                              netns=self.netns,
                              interface=self.interface)
        InterfaceServiceClient(
            "localhost").DeleteInterface(param)
        _, netns = os.path.split(self.netns)
        if len(netns) > 0:
            subprocess.run(["ip", "netns", "delete", netns])
        exit(0)

    def do_get(self):
        logger.info("CNI get is not implemented")
        print("")
        exit(0)

    def do_version(self):
        val, status = json.dumps({'cniVersion': '0.3.1', "supportedVersions": [
            "0.2.0", "0.3.0", "0.3.1"]}), 0
        logger.info("server's version is {}".format(val))
        print(val)
        exit(status)


cni = Cni()
cni.run()

File: mizarcni/main/mizar-cni.go
/*
Copyright 2021 The Mizar Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"bufio"
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"net"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"

	//interface_pb "mizar-cni/client"

	"github.com/containernetworking/cni/pkg/skel"
	"github.com/containernetworking/cni/pkg/types"
	cniTypesVer "github.com/containernetworking/cni/pkg/types/current"
	"github.com/containernetworking/cni/pkg/version"
	"google.golang.org/grpc"
	klog "k8s.io/klog/v2"
)

const (
	NetNSFolder = "/var/run/netns/"
)

type NetVariables struct {
	command         string
	containerID     string
	netNS           string
	ifName          string
	cniPath         string
	k8sPodNamespace string
	k8sPodName      string
	k8sPodTenant    string
	cniVersion      string
	networkName     string
	plugin          string
}

var netVariables NetVariables
var podId PodId
var interfaceId InterfaceId

func init() {
	// Ensures runs only on main thread
	runtime.LockOSThread()

	klog.InitFlags(nil)
	flag.Set("logtostderr", "false")
	flag.Set("log_file", "/tmp/mizarcni.log")

	loadEnvVariables()
	podId = PodId{
		K8SNamespace: netVariables.k8sPodNamespace,
		K8SPodName:   netVariables.k8sPodName,
		K8SPodTenant: netVariables.k8sPodTenant,
	}
	interfaceId = InterfaceId{
		PodId:     &podId,
		Interface: netVariables.ifName,
	}
}

func cmdAdd(args *skel.CmdArgs) error {
	loadNetConf(args.StdinData)
	klog.Infof("Network variables: %s", netVariables)

	// Construct a CniParameters grpc message
	param := CniParameters{
		PodId:     &podId,
		Netns:     netVariables.netNS,
		Interface: netVariables.ifName,
	}
	klog.Infof("Doing CNI add for %s/%s", podId.K8SNamespace, podId.K8SPodName)
	// conn, err := grpc.Dial("localhost:50051", grpc.WithInsecure(), grpc.WithBlock())
	// if err != nil {
	// 	return err
	// }
	// client := NewInterfaceServiceClient(conn)
	// ctx, cancel := context.WithTimeout(context.Background(), time.Second*3)
	client, conn, ctx, cancel, err := getInterfaceServiceClient()
	if err != nil {
		return err
	}
	defer conn.Close()
	defer cancel()
	clientResult, err := client.ConsumeInterfaces(ctx, &param)
	if err != nil {
		return err
	}
	interfaces := clientResult.Interfaces
	klog.Infof("hochan interfaces: %s", interfaces)

	if len(interfaces) == 0 {
		klog.Fatalf("No interfaces found for %s/%s", podId.K8SNamespace, podId.K8SPodName)
	}

	result := cniTypesVer.Result{
		CNIVersion: netVariables.cniVersion,
	}
	for index, interfaceItem := range interfaces {
		result.Interfaces = append(result.Interfaces, &cniTypesVer.Interface{
			Name:    interfaceItem.InterfaceId.Interface,
			Mac:     interfaceItem.Address.Mac,
			Sandbox: netVariables.netNS,
		})

		_, ipnet, err := net.ParseCIDR(fmt.Sprintf("%s/%s", interfaceItem.Address.IpAddress, interfaceItem.Address.IpPrefix))
		if err != nil {
			klog.Info("hochan error")
			return err
		}
		result.IPs = append(result.IPs, &cniTypesVer.IPConfig{
			Version:   interfaceItem.Address.Version,
			Address:   *ipnet,
			Gateway:   net.ParseIP(interfaceItem.Address.GatewayIp),
			Interface: cniTypesVer.Int(index),
		})
	}
	klog.Infof("hochan good: %s", result)
	return result.Print()
}

func cmdDel(args *skel.CmdArgs) error {
	loadNetConf(args.StdinData)
	klog.Infof("Network variables: %s", netVariables)

	result := cniTypesVer.Result{}
	return result.Print()
}

func getInterfaceServiceClient() (InterfaceServiceClient, *grpc.ClientConn, context.Context, context.CancelFunc, error) {
	conn, err := grpc.Dial("localhost:50051", grpc.WithInsecure(), grpc.WithBlock())
	if err != nil {
		return nil, nil, nil, nil, err
	}
	client := NewInterfaceServiceClient(conn)
	ctx, cancel := context.WithTimeout(context.Background(), time.Second*3)
	return client, conn, ctx, cancel, nil
}

func loadNetConf(bytes []byte) {
	netConf := &types.NetConf{}
	if err := json.Unmarshal(bytes, netConf); err != nil {
		if err != nil {
			klog.Fatalf("failed to load netconf: %v", err)
		}
	}

	netVariables.cniVersion = netConf.CNIVersion
	netVariables.networkName = netConf.Name
	netVariables.plugin = netConf.Type
}

func loadEnvVariables() {
	netVariables.command = os.Getenv("CNI_COMMAND")
	netVariables.containerID = os.Getenv("CNI_CONTAINERID")
	netVariables.ifName = os.Getenv("CNI_IFNAME")
	netVariables.cniPath = os.Getenv("CNI_PATH")
	netVariables.netNS = mountNetNSIfNeeded(os.Getenv("CNI_NETNS"))

	cniArgs := os.Getenv("CNI_ARGS")
	if len(cniArgs) > 0 {
		splitted := strings.Split(cniArgs, ";")
		for _, item := range splitted {
			keyValue := strings.Split(item, "=")
			switch keyValue[0] {
			case "K8S_POD_NAMESPACE":
				netVariables.k8sPodNamespace = keyValue[1]
			case "K8S_POD_NAME":
				netVariables.k8sPodName = keyValue[1]
			case "K8S_POD_TENANT":
				netVariables.k8sPodTenant = keyValue[1]
			}
		}
	}
}

func mountNetNSIfNeeded(netNS string) string {
	if !strings.HasPrefix(netNS, NetNSFolder) {
		dstNetNS := strings.ReplaceAll(netNS, "/", "_")
		dstNetNSPath := filepath.Join(NetNSFolder, dstNetNS)
		if netVariables.command == "ADD" {
			os.Mkdir(NetNSFolder, os.ModePerm)
			os.Mkdir(dstNetNSPath, os.ModePerm)
			cmd := exec.Command("mount", "--bind", netNS, dstNetNSPath)
			stderr, _ := cmd.StderrPipe()
			cmd.Start()
			scanner := bufio.NewScanner(stderr)
			var strBuilder strings.Builder
			for scanner.Scan() {
				strBuilder.WriteString(scanner.Text())
			}
			err := cmd.Wait()
			if err != nil {
				klog.Fatalf("failed to bind mount %s to %s: error code %s: %s", netNS, dstNetNSPath, err, strBuilder.String())
			}
		}
		netNS = dstNetNSPath
	}
	return netNS
}

func main() {
	defer klog.Flush()
	skel.PluginMain(cmdAdd, nil, cmdDel, version.PluginSupports("0.2.0", "0.3.0", "0.3.1"), "Mizar CNI plugin")
}
